### DISCUSSION TITLE
Are there any plans for WasmGC integration?

### DISCUSSION THREAD

**Discussion by arximboldi** [emscripten-core/emscripten#23273] [Q&A]
There is a whole set of interesting use-cases that open with WasmGC: https://github.com/WebAssembly/gc

It seems most major browsers support it now.

For example, I develop applications where complex data-models are modeled using simple types combined with immutable data-structures (via Immer: https://github.com/arximboldi/immer). One limitation of this approach is that operating on these data-types from JS is too cumbersome because of the need for `delete()` which goes completely against the expectations of JS, particularly when operating with immutable data. 

For now, I am working around this by having C++ own all the data and expose only references (see https://github.com/emscripten-core/emscripten/discussions/23271). I have also considered just letting those objects "leak". Since they only own memory, it is not strictly a leak if the object is never finalized because it was never collected. The problem with this approach is that because the JS GC does not see into the C++ heap, it may not collect objects from JS even if the C++ heap is getting full.

What would be most interesting is to leverage the JS GC heap directly for all these simple Immer based value trees (Immer does support GC's, e.g Boehm's GC). This would have multiple advantages, even beyond

Are there any plans to expose WasmGC to C or C++? I am also considering that maybe this doesn't even need to be done via Emscripten, but I can do it myself directly. Any of the devs here can give me some hints on how to do this?

**Comment by hoodmane** [comment]
You might want to look at the previous work to allow using `externref` in C/C++:
https://discourse.llvm.org/t/rfc-webassembly-reference-types-in-clang/66939/1

> ### Future
>
> We are currently working on the Garbage Collection WebAssembly proposal which added further reference types, besides externref and funcref. These will be host managed / garbage collected types. This will generalize the implementation of externref and funcref, and we see this patch as a step in the right direction. A future RFC will discuss generalizing the current RFC to handle GC types.

**Reply by hoodmane** [reply]
Here @pmatos says that the igalia funding for this dried up:
https://discourse.llvm.org/t/wasmgc-implementation-status/74821

**Reply by pmatos** [reply]
> Here @pmatos says that the igalia funding for this dried up:
> https://discourse.llvm.org/t/wasmgc-implementation-status/74821

That's correct and still the case. There has been no interest in funding work in this area and therefore we couldn't continue the work which is very unfortunate as this was - at least for me - actually quite exciting to work on.
